src/plugins/GLesmos/shaders.ts

```diff
@@ -0,0 +1,423 @@
+  throw Error(`[GLesmos Error] ${msg}`);
+}
+
+// NOTE: glesmos.replacements:212 must reflect any changes to this type, or you will get errors
+// (Replacement "Replace quadtree implicit tracing with glesmos compilation")
+export interface GLesmosShaderPackage {
+  deps: string[];
+  chunks: GLesmosShaderChunk[];
+}
+
+export interface GLesmosShaderChunk {
+  def: string;
+  color: string;
+  line_color: string;
+  line_width: number;
+}
+
+// I introduced this to make things uniforms more type-safe
+export type GLesmosProgram = WebGLProgram & {
+  vertexAttribPos: number;
+  corner: WebGLUniformLocation | null;
+  size: WebGLUniformLocation | null;
+  NaN: WebGLUniformLocation | null;
+  Infinity: WebGLUniformLocation | null;
+};
+
+type UniformType = "1f" | "2fv" | "3fv" | "4fv" | "1i"; // TODO: this isn't very typesafe!
+export function setUniform(
+  gl: WebGL2RenderingContext,
+  program: WebGLProgram,
+  uniformName: string,
+  uniformType: UniformType,
+  uniformValue: number | number[]
+) {
+  const uniformSetterKey: keyof WebGLRenderingContext = ("uniform" +
+    uniformType) as keyof WebGLRenderingContext;
+  (gl[uniformSetterKey] as Function)(
+    gl.getUniformLocation(program, uniformName),
+    uniformValue
+  );
+}
+
+function compileShader(
+  gl: WebGL2RenderingContext,
+  shaderCode: string,
+  type: number
+) {
+  const shader: WebGLShader | null = gl.createShader(type);
+  if (shader === null) {
+    glesmosError("Invalid shader type");
+  }
+
+  gl.shaderSource(shader, shaderCode);
+  gl.compileShader(shader);
+
+  if (!gl.getShaderParameter(shader, gl.COMPILE_STATUS)) {
+    const shaderInfoLog = gl.getShaderInfoLog(shader);
+    glesmosError(
+      `While compiling ${
+        type === gl.VERTEX_SHADER ? "vertex" : "fragment"
+      } shader:
+      ${shaderInfoLog ?? ""}`
+    );
+  }
+  return shader;
+}
+
+function buildShaderProgram(
+  gl: WebGL2RenderingContext,
+  vert: string,
+  frag: string,
+  _id: string
+) {
+  // console.group('buildShaderProgram');
+  // console.log('vert: %c\n' + vert, 'color: #fc83fc;');
+  // console.log('frag: %c\n' + frag, 'color: #fc83fc;');
+  // console.groupEnd();
+  const shaderProgram = gl.createProgram();
+  if (shaderProgram === null) {
+    glesmosError("Unable to create shader program!");
+  }
+  const vertexShader = compileShader(gl, vert, gl.VERTEX_SHADER);
+  const fragmentShader = compileShader(gl, frag, gl.FRAGMENT_SHADER);
+  if (vertexShader && fragmentShader) {
+    gl.attachShader(shaderProgram, vertexShader);
+    gl.attachShader(shaderProgram, fragmentShader);
+    gl.linkProgram(shaderProgram);
+    return shaderProgram;
+  } else {
+    glesmosError("One or more shaders did not compile.");
+  }
+}
+
+const shaderCache = new Map<string, GLesmosProgram>();
+function getShaderProgram(
+  gl: WebGL2RenderingContext,
+  id: string,
+  vertexSource: string,
+  fragmentSource: string
+) {
+  const key = vertexSource + fragmentSource; // TODO: this is terrible
```

Comment:
I'm not sure what's going on here. Why is this terrible? Do you have an idea for an improvement?

Reply:
Using huge strings as map keys just feels bad.  Hashing the entire shader just to place it in a map can't be _that_ cheap.

That said, adding `vertexShader` probably isn't needed since the same one is used everywhere.